app/src/main/java/com/rkm/tasky/sync/manager/implementation/SyncManagerImpl.kt

```diff
@@ -0,0 +1,124 @@
+import com.rkm.tasky.util.result.onSuccess
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.withContext
+import javax.inject.Inject
+
+class SyncManagerImpl @Inject constructor(
+    private val syncDataSource: SyncDao,
+    private val agendaRemoteDataSource: TaskyAgendaRemoteDataSource,
+    private val eventRemoteDataSource: TaskyEventRemoteDataSource,
+    private val reminderRemoteDataSource: TaskyReminderRemoteDataSource,
+    private val taskRemoteDataSource: TaskyTaskRemoteDataSource,
+    private val createEventRepository: SyncCreateEventRepository,
+    private val updateEventRepository: SyncUpdateEventRepository,
+    private val taskLocalDataSource: TaskDao,
+    private val reminderLocalDataSource: ReminderDao,
+    @IoDispatcher private val dispatcher: CoroutineDispatcher
+): SyncManager {
+    override suspend fun syncDeletedItems(): EmptyResult<NetworkError.APIError> = withContext(dispatcher) {
+        val syncItems = syncDataSource.getSyncItemsByAction(SyncUserAction.DELETE.name)
+        if(syncItems.isNullOrEmpty()) {
+            return@withContext Result.Success(syncItems).asEmptyDataResult()
+        }
+        val networkResult = safeCall { agendaRemoteDataSource.syncAgenda(syncItems.asSyncAgendaRequest()) }
+
+        networkResult.onSuccess {
+            syncDataSource.deleteSyncItems(syncItems.map { item -> item.itemId})
+        }
+
+        return@withContext networkResult.asEmptyDataResult()
+    }
+
+    override suspend fun syncUpdatedItem(): EmptyResult<NetworkError.APIError> = withContext(dispatcher) {
+        val syncItems = syncDataSource.getSyncItemsByAction(SyncUserAction.UPDATE.name)
+        if(syncItems.isNullOrEmpty()) {
+            return@withContext Result.Success(syncItems).asEmptyDataResult()
+        }
+        val taskIds = syncItems.filter { it.item == SyncItemType.TASK }.map { it.itemId }
+        val reminderIds = syncItems.filter { it.item == SyncItemType.REMINDER }.map { it.itemId }
+        val eventsIds = syncItems.filter { it.item == SyncItemType.EVENT }.map { it.itemId }
+
+        val taskToSync = taskLocalDataSource.getTasksById(taskIds)
+        val reminderToSync = reminderLocalDataSource.getRemindersById(reminderIds)
+        val eventsToSync = updateEventRepository.getEvents(eventsIds)
+
+        val itemsToRemoveList = mutableListOf<String>()
+        taskToSync?.forEach { task ->
+            val result = safeCall { taskRemoteDataSource.updateTask(task.asTaskRequest()) }
+            result.onSuccess { itemsToRemoveList.add(task.id) }
+        }
+        reminderToSync?.forEach { reminder ->
+            val result = safeCall { reminderRemoteDataSource.updateReminder(reminder.asReminderRequest()) }
+            result.onSuccess { itemsToRemoveList.add(reminder.id) }
+        }
+        val clearEventList = mutableListOf<String>()
+        eventsToSync.onSuccess { events ->
+            events.forEach {
+                //Implement logic to convert photos
```

Comment:
In this function, the database gets accessed multiple times in a synchronous manner. So before you can schedule the events sync tasks, first the tasks and the reminders need to complete. To me, that does not make much sense - why then have these operations depend on each other? 🤔 

In a situation like this, we can use Kotlin's `async` coroutine builder to have all our operations happen asynchronously. Check out this sample:

```
override suspend fun syncPendingAgendaItems() {
  withContext(Dispatchers.IO) {
    val userId = authInfoStorage.get()?.userId ?: return@withContext
    
    val createdEvents = async {
      eventSyncDao.getAllEventPendingSyncs(
        syncType = SyncType.CREATE,
        userId = userId
      )
    }
    
    val deletedEvents = async {
      eventDeleteSyncDao.getAllEventDeletePendingSyncs(userId)
    }
    
    // .. other sync calls
    
    val createEventJobs = createdEvents.await().map { pendingEntity ->
      launch {
        val event = pendingEntity.event.toDomain() // convert to domain model Event
        
        // try to make the API call immediately
        when(remoteDataSource.postEvent(event)) {
          is Result.Error -> Unit
          is Result.Success -> applicationScope.launch {
            pendingSyncDao.deleteEventPendingSyncEntity(event.id)
          }.join()
        }
      }
    }
    
    // ... other API calls
    
    // Then we await the completion of each before the function terminates
    createEventJobs.forEach { it.join() }
    deleteEventJobs.forEach { it.join() }
    createTaskJobs.forEach { it.join() }
    deletedTaskJobs.forEach { it.join() }
    // ...    
  }
}
```